File: examples/python/managed_identity_example.py
"""
Azure Managed Redis - Managed Identity Authentication Example

This example demonstrates how to connect to Azure Managed Redis using
a User-Assigned Managed Identity with Entra ID authentication.

Requirements:
- redis>=5.0.0
- redis-entraid>=1.0.0
- python-dotenv>=1.0.0 (optional, for .env file support)

Environment Variables:
- AZURE_CLIENT_ID: Client ID of the user-assigned managed identity
- REDIS_HOSTNAME: Hostname of your Azure Managed Redis instance
- REDIS_PORT: Port (default: 10000)

This code should be run from an Azure resource (App Service, VM, etc.)
that has the managed identity assigned.
"""

import os
import sys
import logging
from datetime import datetime

# Try to load .env file if present (useful for local testing with service principal)
try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass

from redis import Redis, RedisError
from redis_entraid.cred_provider import (
    create_from_managed_identity,
    ManagedIdentityType,
    ManagedIdentityIdType
)

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


def get_config():
    """Load configuration from environment variables."""
    config = {
        'client_id': os.environ.get('AZURE_CLIENT_ID'),
        'redis_host': os.environ.get('REDIS_HOSTNAME'),
        'redis_port': int(os.environ.get('REDIS_PORT', 10000)),
    }
    
    # Validate required config
    if not config['client_id']:
        raise ValueError("AZURE_CLIENT_ID environment variable is required")
    if not config['redis_host']:
        raise ValueError("REDIS_HOSTNAME environment variable is required")
    
    return config


def create_redis_client(config: dict) -> Redis:
    """
    Create a Redis client with Entra ID authentication using managed identity.
    
    Args:
        config: Configuration dictionary with client_id, redis_host, redis_port
    
    Returns:
        Redis client instance
    """
    logger.info(f"Creating credential provider for managed identity: {config['client_id'][:8]}...")
    
    # Create credential provider for user-assigned managed identity
    credential_provider = create_from_managed_identity(
        identity_type=ManagedIdentityType.USER_ASSIGNED,
        resource="https://redis.azure.com/",
        id_type=ManagedIdentityIdType.CLIENT_ID,
        id_value=config['client_id']
    )
    
    logger.info(f"Connecting to Redis at {config['redis_host']}:{config['redis_port']}")
    
    # Create Redis client
    client = Redis(
        host=config['redis_host'],
        port=config['redis_port'],
        credential_provider=credential_provider,
        ssl=True,
        decode_responses=True,
        socket_connect_timeout=10,
        socket_timeout=10
    )
    
    return client


def test_connection(client: Redis) -> bool:
    """Test the Redis connection with a ping."""
    try:
        result = client.ping()
        logger.info(f"PING response: {result}")
        return result
    except RedisError as e:
        logger.error(f"PING failed: {e}")
        return False


def run_demo_operations(client: Redis):
    """Run demonstration Redis operations."""
    print("\n" + "="*60)
    print("AZURE MANAGED REDIS - ENTRA ID AUTHENTICATION DEMO")
    print("="*60 + "\n")
    
    # Test 1: Basic PING
    print("1. Testing connection with PING...")
    if test_connection(client):
        print("   ✅ Connection successful!\n")
    else:
        print("   ❌ Connection failed!\n")
        return
    
    # Test 2: SET operation
    print("2. Testing SET operation...")
    test_key = f"entra-auth-test:{datetime.now().isoformat()}"
    test_value = "Hello from Entra ID authenticated client!"
    try:
        client.set(test_key, test_value, ex=60)  # Expires in 60 seconds
        print(f"   ✅ SET '{test_key}' = '{test_value}'\n")
    except RedisError as e:
        print(f"   ❌ SET failed: {e}\n")
        return
    
    # Test 3: GET operation
    print("3. Testing GET operation...")
    try:
        retrieved = client.get(test_key)
        print(f"   ✅ GET '{test_key}' = '{retrieved}'\n")
    except RedisError as e:
        print(f"   ❌ GET failed: {e}\n")
        return
    
    # Test 4: INFO command
    print("4. Getting server info...")
    try:
        info = client.info('server')
        print(f"   Redis Version: {info.get('redis_version', 'N/A')}")
        print(f"   Redis Mode: {info.get('redis_mode', 'N/A')}\n")
    except RedisError as e:
        print(f"   ❌ INFO failed: {e}\n")
    
    # Test 5: Delete test key
    print("5. Cleaning up test key...")
    try:
        client.delete(test_key)
        print(f"   ✅ Deleted '{test_key}'\n")
    except RedisError as e:
        print(f"   ❌ DELETE failed: {e}\n")
    
    print("="*60)
    print("DEMO COMPLETE - All operations successful!")
    print("="*60)


def main():
    """Main entry point."""
    try:
        # Load configuration
        config = get_config()
        
        # Create Redis client
        client = create_redis_client(config)
        
        # Run demo operations
        run_demo_operations(client)
        
        # Close connection
        client.close()
        
    except ValueError as e:
        logger.error(f"Configuration error: {e}")
        sys.exit(1)
    except RedisError as e:
        logger.error(f"Redis error: {e}")
        sys.exit(1)
    except Exception as e:
        logger.error(f"Unexpected error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: examples/nodejs/managed_identity_example.mjs
/**
 * Azure Managed Redis - Managed Identity Authentication Example (Node.js)
 * 
 * This example demonstrates how to connect to Azure Managed Redis using
 * a User-Assigned Managed Identity with Entra ID authentication.
 * 
 * Requirements:
 * - Node.js 18+
 * - redis (node-redis)
 * - @azure/identity
 * 
 * Environment Variables:
 * - AZURE_CLIENT_ID: Client ID of the user-assigned managed identity
 * - REDIS_HOSTNAME: Hostname of your Azure Managed Redis instance
 * - REDIS_PORT: Port (default: 10000)
 * 
 * This code should be run from an Azure resource that has the
 * managed identity assigned.
 */

import { createClient } from 'redis';
import { ManagedIdentityCredential } from '@azure/identity';

// Load configuration from environment
const config = {
  clientId: process.env.AZURE_CLIENT_ID,
  principalId: process.env.PRINCIPAL_ID || '', // Object ID for AUTH username
  redisHost: process.env.REDIS_HOSTNAME,
  redisPort: parseInt(process.env.REDIS_PORT || '10000', 10)
};

// Redis scope for Entra ID
const REDIS_SCOPE = 'https://redis.azure.com/.default';

// Validate configuration
function validateConfig() {
  const missing = [];
  if (!config.clientId) missing.push('AZURE_CLIENT_ID');
  if (!config.redisHost) missing.push('REDIS_HOSTNAME');
  
  if (missing.length > 0) {
    console.error(`Error: Missing required environment variables: ${missing.join(', ')}`);
    console.error('\nPlease set:');
    console.error("  export AZURE_CLIENT_ID='your-managed-identity-client-id'");
    console.error("  export REDIS_HOSTNAME='your-redis.region.redis.azure.net'");
    process.exit(1);
  }
}

// Extract OID from JWT token for use as username
function extractOidFromToken(token) {
  try {
    const parts = token.split('.');
    if (parts.length !== 3) return null;
    const payload = JSON.parse(Buffer.from(parts[1], 'base64').toString());
    return payload.oid;
  } catch {
    return null;
  }
}

async function main() {
  validateConfig();

  console.log('\n' + '='.repeat(60));
  console.log('AZURE MANAGED REDIS - NODE.JS MANAGED IDENTITY AUTH DEMO');
  console.log('='.repeat(60) + '\n');

  let client;
  
  try {
    // Create managed identity credential
    console.log('1. Creating managed identity credential...');
    const credential = new ManagedIdentityCredential(config.clientId);
    console.log(`   ✅ Credential created for: ${config.clientId.substring(0, 8)}...\n`);

    // Get initial token to extract OID for username
    console.log('2. Acquiring initial token...');
    const tokenResponse = await credential.getToken(REDIS_SCOPE);
    if (!tokenResponse) {
      throw new Error('Failed to acquire token');
    }
    const oid = extractOidFromToken(tokenResponse.token);
    console.log(`   ✅ Token acquired, OID: ${oid || 'unknown'}`);
    console.log(`   Token expires: ${new Date(tokenResponse.expiresOnTimestamp).toISOString()}\n`);

    // Create Redis client with Entra ID auth
    console.log('3. Creating Redis client...');
    const redisUrl = `rediss://${config.redisHost}:${config.redisPort}`;
    
    client = createClient({
      url: redisUrl,
      username: oid || config.principalId,
      password: tokenResponse.token,
      socket: {
        tls: true,
        servername: config.redisHost
      }
    });

    // Handle errors
    client.on('error', (err) => console.error('Redis Client Error:', err));
    
    console.log(`   ✅ Client configured for ${redisUrl}\n`);

    // Connect
    console.log('4. Connecting to Redis...');
    await client.connect();
    console.log('   ✅ Connected!\n');

    // Test PING
    console.log('5. Testing PING...');
    const pong = await client.ping();
    console.log(`   ✅ PING response: ${pong}\n`);

    // Test SET
    console.log('6. Testing SET operation...');
    const testKey = `nodejs-entra-test:${new Date().toISOString()}`;
    const testValue = 'Hello from Node.js with Entra ID auth!';
    await client.set(testKey, testValue, { EX: 60 }); // Expires in 60 seconds
    console.log(`   ✅ SET '${testKey}'\n`);

    // Test GET
    console.log('7. Testing GET operation...');
    const retrieved = await client.get(testKey);
    console.log(`   ✅ GET '${testKey}' = '${retrieved}'\n`);

    // Server info
    console.log('8. Getting server info...');
    const info = await client.info('server');
    const versionMatch = info.match(/redis_version:(.+)/);
    console.log(`   Redis Version: ${versionMatch ? versionMatch[1].trim() : 'unknown'}\n`);

    // Clean up
    console.log('9. Cleaning up test key...');
    await client.del(testKey);
    console.log(`   ✅ Deleted '${testKey}'\n`);

    console.log('='.repeat(60));
    console.log('DEMO COMPLETE - All operations successful!');
    console.log('='.repeat(60) + '\n');

  } catch (error) {
    console.error('\n❌ Error:', error.message);
    if (error.code) {
      console.error(`   Error code: ${error.code}`);
    }
    console.error('\nTroubleshooting tips:');
    console.error('1. Ensure you are running on an Azure resource with managed identity');
    console.error('2. Verify the managed identity has an access policy on the Redis cache');
    console.error('3. Check that AZURE_CLIENT_ID matches your user-assigned managed identity');
    process.exit(1);
  } finally {
    if (client) {
      await client.quit();
    }
  }
}

main();
